model.py

```diff
@@ -249,33 +267,54 @@ def decode_mae(self, x, ids_restore):
         x = x[:, 1:, :]
         return x
 
+    def encode_causal(self, x, mask_ratio):
+        x = self.conv1(x)
+        x = x.reshape((x.shape[0], -1, x.shape[-1]))
+        T = x.shape[1]
+        x = x + self.positional_embedding[1:, :].astype(x.dtype)
+
+        len_keep = int(x.shape[1] * (1 - mask_ratio))
+        x = x[:, :len_keep]
+
+        x = self.fwd_transformer(x)
+        return x, T
+
+    def decode_causal(self, x, T):
+        x = self.decoder_embed(x)
+        mask_tokens = jnp.tile(
+            self.mask_token,
+            (x.shape[0], T - x.shape[1] + 1, 1)
+        )
+        x = jnp.concatenate([x, mask_tokens], axis=1)
+
+        x = x + self.decoder_positional_embedding.astype(x.dtype)
+        x = self.dec_ln_pre(x)
+        attn_mask = nn.make_causal_mask(jnp.ones((x.shape[0], T + 1), dtype=x.dtype))
```

Comment:
Nice! So the encoder is non-causal and the decoder is causal, so it's basically a PrefixLM style thing. Could be really good for pretraining because the decoder sees way more data samples (parallelizes across the sequence for next-patch prediction) rather than 15% of the masked examples.